lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -2143,51 +2143,261 @@ namespace Js
+        UNumberFormatFields *fields;
+
+        static const UNumberFormatFields UnsetField = static_cast<UNumberFormatFields>(0xFFFFFFFF);
+
+        JavascriptString *GetPartTypeString(UNumberFormatFields field)
+        {
+            JavascriptLibrary *library = sc->GetLibrary();
+
+            if (field == UnsetField)
+            {
+                return library->GetIntlLiteralPartString();
+            }
+
+            switch (field)
+            {
+            case UNUM_INTEGER_FIELD:
+            {
+                if (JavascriptNumber::IsNan(num))
+                {
+                    return JavascriptString::NewCopySz(_u("nan"), sc);
+                }
+                else if (!NumberUtilities::IsFinite(num))
+                {
+                    return JavascriptString::NewCopySz(_u("infinity"), sc);
+                }
+                else
+                {
+                    return JavascriptString::NewCopySz(_u("integer"), sc);
+                }
+            }
+            case UNUM_FRACTION_FIELD: return JavascriptString::NewCopySz(_u("fraction"), sc);
+            case UNUM_DECIMAL_SEPARATOR_FIELD: return JavascriptString::NewCopySz(_u("decimal"), sc);
+
+            // The following three should only show up if UNUM_SCIENTIFIC is used, which Intl.NumberFormat doesn't currently use
+            case UNUM_EXPONENT_SYMBOL_FIELD: AssertOrFailFastMsg(false, "Unexpected exponent symbol field");
+            case UNUM_EXPONENT_SIGN_FIELD: AssertOrFailFastMsg(false, "Unexpected exponent sign field");
+            case UNUM_EXPONENT_FIELD: AssertOrFailFastMsg(false, "Unexpected exponent field");
+
+            case UNUM_GROUPING_SEPARATOR_FIELD: return JavascriptString::NewCopySz(_u("group"), sc);
+            case UNUM_CURRENCY_FIELD: return JavascriptString::NewCopySz(_u("currency"), sc);
+            case UNUM_PERCENT_FIELD: return JavascriptString::NewCopySz(_u("percent"), sc);
+
+            // TODO(jahorto): Determine if this would ever be returned and what it would map to
+            case UNUM_PERMILL_FIELD: AssertOrFailFastMsg(false, "Unexpected permill field");
+
+            case UNUM_SIGN_FIELD: num < 0 ? JavascriptString::NewCopySz(_u("minusSign"), sc) : JavascriptString::NewCopySz(_u("plusSign"), sc);
+            default: AssertOrFailFastMsg(false, "Unexpected unknown part"); return nullptr;
+            }
+        }
+
+    public:
+        NumberFormatPartsBuilder(double num, const char16 *formatted, charcount_t formattedLength, ScriptContext *scriptContext)
+            : num(num)
+            , formatted(formatted)
+            , formattedLength(formattedLength)
+            , sc(scriptContext)
+            , fields(RecyclerNewArrayLeaf(sc->GetRecycler(), UNumberFormatFields, formattedLength))
+        {
+            // this will allow InsertPart to tell what fields have been initialized or not, and will
+            // be the valid of resulting { type: "literal" } fields in ToPartsArray
```

Comment:
>be the valid of resulting [](start = 15, length = 25)

what?

Reply:
I think I meant "the *value* of resulting...". Does that make more sense?